qiskit_optimization/algorithms/qrao/quantum_random_access_encoding.py

```diff
@@ -0,0 +1,570 @@
+    qrac_circ = reduce(lambda x, y: x.tensor(y), qracs)
+    return qrac_circ
+
+
+class QuantumRandomAccessEncoding:
+    """This class specifies a Quantum Random Access Code that can be used to encode
+    the binary variables of a QUBO (quadratic unconstrained binary optimization
+    problem).
+
+    Args:
+        max_vars_per_qubit: maximum possible compression ratio.
+            Supported values are 1, 2, or 3.
+
+    """
+
+    # This defines the convention of the Pauli operators (and their ordering)
+    # for each encoding.
+    OPERATORS = (
+        (SparsePauliOp("Z"),),  # (1,1,1) QRAC
+        (SparsePauliOp("X"), SparsePauliOp("Z")),  # (2,1,p) QRAC, p ≈ 0.85
+        (SparsePauliOp("X"), SparsePauliOp("Y"), SparsePauliOp("Z")),  # (3,1,p) QRAC, p ≈ 0.79
+    )
+
+    def __init__(self, max_vars_per_qubit: int = 3):
+        if max_vars_per_qubit not in (1, 2, 3):
+            raise ValueError("max_vars_per_qubit must be 1, 2, or 3")
+        self._ops = self.OPERATORS[max_vars_per_qubit - 1]
+
+        self._qubit_op: Optional[SparsePauliOp] = None
+        self._offset: Optional[float] = None
+        self._problem: Optional[QuadraticProgram] = None
+        self._var2op: Dict[int, Tuple[int, SparsePauliOp]] = {}
+        self._q2vars: List[List[int]] = []
+        self._frozen = False
+
+    @property
+    def num_qubits(self) -> int:
+        """Number of qubits"""
+        return len(self._q2vars)
+
+    @property
+    def num_vars(self) -> int:
+        """Number of decision variables"""
+        return len(self._var2op)
+
+    @property
+    def max_vars_per_qubit(self) -> int:
+        """Maximum number of variables per qubit"""
+        return len(self._ops)
+
+    @property
+    def var2op(self) -> Dict[int, Tuple[int, SparsePauliOp]]:
+        """Maps each decision variable to ``(qubit_index, operator)``"""
+        return self._var2op
+
+    @property
+    def q2vars(self) -> List[List[int]]:
+        """Each element contains the list of decision variable indices encoded on that qubit"""
+        return self._q2vars
+
+    @property
+    def compression_ratio(self) -> float:
+        """Compression ratio. Number of decision variables divided by number of qubits"""
+        return self.num_vars / self.num_qubits
+
+    @property
+    def minimum_recovery_probability(self) -> float:
+        """Minimum recovery probability, as set by ``max_vars_per_qubit``"""
+        n = self.max_vars_per_qubit
+        return (1 + 1 / np.sqrt(n)) / 2
+
+    @property
+    def qubit_op(self) -> SparsePauliOp:
+        """Relaxed Hamiltonian operator.
+
+        Raises:
+            AttributeError: If no objective function has been provided yet, and
+                a qubit Hamiltonian cannot be constructed. Use the `encode` method
+                to manually compile this field.
+        """
+        if self._qubit_op is None:
+            raise AttributeError(
```

Comment:
I am not sure AttributeError is the right thing. Maybe just a general RuntimeError or the QiskitOptimizationError.

Reply:
I chose `AttributeError` because an error here literally means that the attribute/property has not yet been set.  Similar to

```python
>>> class A():
...     pass
... 
>>> a = A()
>>> a.b
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'A' object has no attribute 'b'
```

but `RuntimeError` or similar is not a bad choice -- and may in fact be a better one -- because this error indicates that you have accessed these without having first called `encode`.